src/lib/input/input-wrapper.ts

```diff
@@ -0,0 +1,289 @@
+@Directive({
+  selector: 'md-hint',
+  host: {
+    '[class.md-right]': 'align == "end"',
+    '[class.md-hint]': 'true'
+  }
+})
+export class MdHint {
+  // Whether to align the hint label at the start or end of the line.
+  @Input() align: 'start' | 'end' = 'start';
+}
+
+
+/**
+ * Component that represents a text input. It encapsulates the <input> HTMLElement and
+ * improve on its behaviour, along with styling it according to the Material Design.
+ */
+@Component({
+  moduleId: module.id,
+  selector: 'md-input-wrapper',
+  templateUrl: 'input-wrapper.html',
+  styleUrls: ['input.css', 'input-wrapper.css'],
+  host: {
+    '(click)' : '_focusInput()',
+    // remove align attribute to prevent it from interfering with layout.
+    '[attr.align]': 'null',
+  },
+  encapsulation: ViewEncapsulation.None,
+})
+export class MdInputWrapper implements AfterContentInit, OnChanges {
+  @Input() align: 'start' | 'end' = 'start';
+
+  @Input() dividerColor: 'primary' | 'accent' | 'warn' = 'primary';
+
+  @Input() hintLabel: string = '';
+
+  @Input()
+  get floatingPlaceholder(): boolean { return this._floatingPlaceholder; }
+  set floatingPlaceholder(value) { this._floatingPlaceholder = coerceBooleanProperty(value); }
+  private _floatingPlaceholder: boolean = true;
+
+  @ContentChild(MdPlaceholder) _placeholderChild: MdPlaceholder;
+
+  @ContentChildren(MdHint) _hintChildren: QueryList<MdHint>;
+
+  @ContentChild(NgModel) _ngModelChild: NgModel;
+
+  /** Whether the `input` or `textarea` is focused. */
+  _focused = false;
+
+  /** The disabled attribute of the `input` or `textarea`. */
+  _inputDisabled = false;
+
+  /** The id attribute of the `input` or `textarea`. */
+  _inputId = '';
+
+  /** The required attribute of the `input` or `textarea`. */
+  _inputRequired = false;
+
+  /** Whether the `input` or `textarea` is empty. */
+  get _empty(): boolean { return !this._inputValue && this._inputType !== 'date'; }
+
+  /** The placeholder attribute of the `input` or `textarea`. */
+  get _inputPlaceholder(): string { return this._getterSetterOnlyInputPlaceholder; }
+  set _inputPlaceholder(value: string) {
+    this._getterSetterOnlyInputPlaceholder = value;
+    this._validatePlaceholders();
+  }
+  private _getterSetterOnlyInputPlaceholder = '';
+
+  /** The type attribute of the `input` (or "text" if element is a `textarea`). */
+  private get _inputType(): string { return this._getterSetterOnlyInputType; }
+  private set _inputType(value: string) {
+    this._getterSetterOnlyInputType = value || 'text';
+    this._validateInputType();
+  }
+  private _getterSetterOnlyInputType = 'text';
+
+  /** The value of the `input` or `textarea`. */
+  private _inputValue = '';
+
+  /** The native `input` or `textarea` element. */
+  private _inputElement: HTMLInputElement | HTMLTextAreaElement;
+
+  /** A `MutationObserver` to observe the `input` or `textarea`. */
+  private _inputObserver = new MutationObserver(mutations => {
```

Comment:
This might make web workers freak out. @jelbourn what's the proper way of handling mutations like those on elements from the content tree? Create a service with an Observable (that would be my guess)?